KAFKA-15534: Inject request completion time when the request failed

philipnee opened the pull request:
Currently, we aren't able to access the request completion time if the request is completed exceptionally, which results in many system calls.  This is not ideal because these system calls can add up.  Instead, time is already retrieved on the top of the background thread event loop, which is then propagated into the NetworkClientDelegate.poll.

In this PR - I store the completion time in the handler, so that it becomes accessible in the callbacks.

philipnee committed:
propagate completion time to the onFailure to get proper backoff

cadonna commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java:
```diff
@@ -290,7 +290,6 @@ private void onFailure(final long currentTimeMs,
 
         private void retry(final long currentTimeMs) {
             onFailedAttempt(currentTimeMs);
-            onSendAttempt(currentTimeMs);
```
Why is this call removed?

philipnee replied:
This is duplicated because the manager only sends out the request when it is being polled via the `poll()` method, which `drain()` the unsent requests from the queue.  In the `drain()` it already invokes the `onSentAttempt()`

See
```
for (OffsetFetchRequestState request : partitionedBySendability.get(true)) {
                request.onSendAttempt(currentTimeMs);
                unsentRequests.add(request.toUnsentRequest());
                inflightOffsetFetches.add(request);
            }
```

cadonna replied:
Ah, yeah that makes sense!

cadonna commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -254,28 +250,38 @@ public String toString() {
+    public static class FutureCompletionHandler extends CompletableFuture<ClientResponse> implements RequestCompletionHandler {
 
-        private final CompletableFuture<ClientResponse> future;
+        /**
+         * The time when the response is completed. This is used when the response is completed exceptionally because
+         * ClientResponse already contains received time which is injected by the network client.
+         */
```
I do not understand this comment. Variable `responseCompletionTimeMs` is also set when the request is completed succesfully. I actually think, we do not need this comment.

philipnee replied:
Thanks, I think I was trying to say the main use of responseCompletionTimeMs is mainly used when the future gets completed exceptionally.  For the successful requests, we could use either the ClientResponse or this time.  I'll remove this comment to avoid the ambiguity.

cadonna commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:
```diff
@@ -184,13 +184,11 @@ private NetworkClientDelegate.UnsentRequest makeHeartbeatRequest() {
-                // TODO: Currently, we lack a good way to propage the response time from the network client to the
-                //  request handler. We will need to store the response time in the handler to make it accessible.
-                onFailure(exception, time.milliseconds());
+                onFailure(exception, request.handler().completionTimeMs());
```
I do not completely understand how this should work. If the response is `null` then I assume also `onComplete()` of `NetworkClientDelegate` is not called because that would lead to a `NullPointerException`. If `onComplete()` of `NetworkClientDelegate` is not called the `completionTimeMs` field in `NetworkClientDelegate` is not set. Thus, `request.handler().completionTimeMs()` will not return the completion time.
Am I missing something?

philipnee replied:
I think the only time we invoke onComplete() is when a response is available (see below). If the response is `null`, which indicates the request has not been sent out; therefore, in the `NetworkClientDelegate`, we need to actively fail the request on timeout.

onComplete when response is available:
```
private void completeResponses(List<ClientResponse> responses) {
        for (ClientResponse response : responses) {
            try {
                response.onComplete();
            } catch (Exception e) {
                log.error("Uncaught error in request completion:", e);
            }
        }
    }
```

We actively expires the unsent request and fail them with TimeoutException:
```
if (unsent.timer.isExpired()) {
                iterator.remove();
                unsent.handler.onFailure(currentTimeMs, new TimeoutException(
                    "Failed to send request after " + unsent.timer.timeoutMs() + " ms."));
                continue;
            }
```

cadonna replied:
I am not sure, I understand. If `onComplete()` is never called with response == `null` why is there a branch in the completion handler for that case that calls `onFailure()`? Wouldn't it be better to throw an `IllegalStateException` or similar?

philipnee replied:
The future may be completed exceptionally with an exception, or get completed with a response and an error code.  Is this where the confusion comes from? As retriable exception can be thrown in both response as well as the throwable part. 

Should I make onComplete to accept both the error and the response? like void onComplete(ClientResponse response, Throwable error); to be more consistent?

cadonna replied:
OK, this was a misunderstanding on my side.

cadonna reviewed (commented):
Thanks for the PR, @philipnee !

I started to review the PR and have a fundamental question in HeartbeatRequestManager.

philipnee commented:
Hi @cadonna - Thank you for putting time into this PR, very much appreciated. I responded to some of your questions, let me know if there is still any ambiguity left.

philipnee committed:
Remove ambiguous comment

cadonna commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -254,28 +250,34 @@ public String toString() {
-    public static class FutureCompletionHandler implements RequestCompletionHandler {
-
-        private final CompletableFuture<ClientResponse> future;
+    public static class FutureCompletionHandler extends CompletableFuture<ClientResponse> implements RequestCompletionHandler {
```
Why do you want to inherit from `CompletableFuture`? Extending base classes might lead to the fragile base class problem. I do not see that your code manifests this problem at the moment, but I think it would be better to minimize the surface for such problem if possible and if it does not cost too much.

philipnee committed:
remove extension to future

revert some changes

remove weird typing

Update ClientResponse.java

wip

wip

revert some changes

cadonna commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManager.java:
```diff
@@ -148,30 +148,33 @@ private Optional<NetworkClientDelegate.UnsentRequest> send(final long currentTim
+
+            unsent.future().whenComplete((response, exception) -> {
+                if (response == null) {
+                    // Backoff if the error is retriable
```
I think this is not the right place for this comment because there is no concept of a backoff here. I think you should remove it.

(Maybe you realised by now that I am not a big fan of inline comments. I think they eventually start to lie. In this case it is even very likely because you cannot see the code to which this comment refers. If something changes in `handleError()`, it is very unlikely that this comment will be updated.)

cadonna commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:
```diff
@@ -188,9 +188,7 @@ private NetworkClientDelegate.UnsentRequest makeHeartbeatRequest() {
             if (response != null) {
                 onResponse((ConsumerGroupHeartbeatResponse) response.responseBody(), response.receivedTimeMs());
```
nit:
I was wondering whether you should use `request.handler().completionTimeMs()` instead of `response.receivedTimeMs()` here and in other places. IMO, it would make the calls clearer because `request.handler()` would be the single component that manages the completion time.
I also realized that if the completion handler is passed into the constructor of `UnsentRequest` using  `request.handler()` would not be possible. I was also thinking that maybe you should get rid of that constructor because it makes the code less readable. 
Those are just nit comments to think about. They do not block merging this PR.

philipnee replied:
Thanks
`request.handler().completionTimeMs() instead of response.receivedTimeMs()` - This is addressed

`UnsentRequest using request.handler() would not be possible` - Let me quickly follow up with a MINOR PR to address this comment.

cadonna commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java:
```diff
@@ -256,27 +256,39 @@ public String toString() {
 
-        public void onFailure(final RuntimeException e) {
-            future.completeExceptionally(e);
+        public void onFailure(final long currentTimeMs, final RuntimeException e) {
```
I could not find unit tests to verify the setting of the correct completion time in `onFailure()` and `onComplete()`.

philipnee replied:
You are right, I should add one there.

cadonna commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java:
```diff
@@ -254,14 +254,14 @@ public NetworkClientDelegate.UnsentRequest toUnsentRequest() {
-                    (r, t) -> onResponse(r.receivedTimeMs(), (OffsetFetchResponse) r.responseBody()));
+                builder,
+                coordinatorRequestManager.coordinator(),
+                (r, t) -> onResponse(r.receivedTimeMs(), (OffsetFetchResponse) r.responseBody()));
```
Out of curiosity, can the response never be `null` here as in `HeartbeatRequestManager`? Can timeouts not happen here?

philipnee replied:
In fact, this is a bug and I'm planning to address this in [KAFKA-15562](https://issues.apache.org/jira/browse/KAFKA-15562).  Can we address the gap all in a follow up PR?

cadonna replied:
Of course, mine was a question for my understanding.

cadonna commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManagerTest.java:
```diff
@@ -151,6 +152,36 @@ void testHardFailures(Exception exception) {
     }
 
+    @Test
+    public void testNetworkTimeout() {
```
I expected such a test also in the other request manager tests. The only request manager test that has such test other than this is `HeartbeatRequestManager`.

philipnee replied:
Per previous comment: NetworkTimeout isn't handled correctly in the commitRequestManager, therefore I want to address this in that PR.  I will check other request manager to ensure this is tested.

philipnee replied:
CoordinatorRequestManager doesn't have it - so I added the test.

philipnee replied:
Not entirely sure about the behavior of the OffsetRequestManager - I'll need to get back to this.  Added a ticket there: KAFKA-15642

cadonna reviewed (commented):
Thanks for the updates and the clarifications, @philipnee !

Here my comments.

philipnee committed:
addressed PR comments!

philipnee commented:
Hi @cadonna Thank you for spending time reviewing the PR again. I have answered your comments, and attached a followup ticket and a minor PR: https://github.com/apache/kafka/pull/14581

The PR is based on this PR so the commit history looks a bit messy; once we merge this PR, I will rebase again.

Let me know if you have more questions around this PR, love to address them.

cadonna commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegateTest.java:
```diff
@@ -97,6 +99,29 @@ public void testTimeoutAfterSend() throws Exception {
+        time.sleep(100);
+        assertEquals(timeMs, unsentRequest.handler().completionTimeMs());
+    }
+
```
If you use a mock for the client response, the test becomes simpler.
```suggestion
@Test
public void testEnsureCorrectCompletionTimeOnComplete() {
    NetworkClientDelegate.UnsentRequest unsentRequest = newUnsentFindCoordinatorRequest();
    long timeMs = time.milliseconds();
    final ClientResponse response = mock(ClientResponse.class);
    when(response.receivedTimeMs()).thenReturn(timeMs);

    unsentRequest.handler().onComplete(response);
    time.sleep(100);

    assertEquals(timeMs, unsentRequest.handler().completionTimeMs());
}
```

cadonna reviewed (approved):
Thanks for the updates, @philipnee!

LGTM!

I have just one minor comment.

philipnee committed:
Update NetworkClientDelegateTest.java

philipnee commented:
Thank you @cadonna - I adapted your suggestion. Thanks again!

philipnee committed:
Update CommitRequestManager.java

trigger commit

cadonna commented:
Build failures are unrelated.
